crates/ruff_linter/src/rules/ruff/rules/spurious_async.rs

```diff
@@ -0,0 +1,83 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::visitor;
+use ruff_python_ast::visitor::Visitor;
+use ruff_python_ast::{self as ast, Expr, Stmt};
+use ruff_text_size::TextRange;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for functions declared `async` that do not yield or otherwise use features requiring the
+/// function to be declared `async`.
+///
+/// ## Why is this bad?
+/// Declaring a function `async` when it's not is usually a mistake, and may artificially limit the
+/// contexts where that function may be called.
+#[violation]
+pub struct SpuriousAsync {
+    name: String,
+}
+
+impl Violation for SpuriousAsync {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let SpuriousAsync { name } = self;
+        format!("Function `{name}` is declared `async`, but doesn't yield or use async features.")
+    }
+}
+
+#[derive(Default)]
+struct YieldingExprVisitor {
+    found_yield: bool,
+}
+
+impl<'a> Visitor<'a> for YieldingExprVisitor {
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        match expr {
+            Expr::Await(_) => {
+                self.found_yield = true;
+            }
+            _ => visitor::walk_expr(self, expr),
+        }
+    }
+    fn visit_stmt(&mut self, stmt: &'a Stmt) {
+        match stmt {
+            Stmt::With(ast::StmtWith { is_async: true, .. }) => {
+                self.found_yield = true;
+            }
+            Stmt::For(ast::StmtFor { is_async: true, .. }) => {
+                self.found_yield = true;
+            }
+            _ => visitor::walk_stmt(self, stmt),
+        }
+    }
+}
+
+/// RUF029
+pub(crate) fn spurious_async(
+    checker: &mut Checker,
+    is_async: bool,
+    name: &str,
+    body: &[Stmt],
+    range: TextRange,
+) {
+    if !is_async {
```

Comment:
This is a Python noob question from a JS engineer. How does Python handle default arguments? Can you use `await` in a default initializer? Does this count as async call (because it pauses the function execution) or not because defaults are initialized once during declaration and not when calling the function.

Reply:
That's a fun question that I'd never considered before.

`await` isn't allowed at the top-level in a usual Python file (only inside async functions), so normally this isn't really something that would ever come up, since function defaults are evaluated in the scope the function is in rather than the scope the function creates:

```pycon
>>> import asyncio
>>> async def foo(x=await asyncio.sleep(0.1)): ...
...
  File "<stdin>", line 1
SyntaxError: 'await' outside function
```

But theoretically you could use `await` in a function default argument if it was an inner function inside an async function:

```pycon
>>> import asyncio
>>> async def foo():
...     def bar(x=await asyncio.sleep(0.1)): ...
...     return bar
...
>>> bar = asyncio.run(foo())
>>> import inspect
>>> inspect.iscoroutinefunction(bar)
False
```

As you can see from that REPL snippet, using `await` inside the default argument for the inner function doesn't impact whether or not the inner function is considered to be a coroutine function, since the `await` in the function's default argument is evaluated before the function has even finished being defined.

Reply:
Thanks for the detailed explanation.  

I tested if `foo` is considered a co-routine and that seems to be the case. Do you know if that's only because the function is marked as `async`? 

```python
>>> import asyncio
>>> async def foo():
...     def bar(x=await asyncio.sleep(0.1)): ...
...     return bar
... 
>>> import inspect
>>> inspect.iscoroutinefunction(foo)
True
```

I'm trying to understand if the use of `async` on `foo` is something this lint rule should warn about or not.

Reply:
`foo` is considered a coroutine function because it returns a coroutine ;)

```pycon
>>> foo
<function foo at 0x00000274CBC2A160>
>>> foo()
<coroutine object foo at 0x00000274CDC19300>
```

But the short answer to your question is "yes" -- all `async def` functions are "coroutine functions" -- functions that return coroutine objects that can then be `await`ed inside other coroutine functions.

The use of `async` on `foo` should not be something that this lint rule should warn about, because the `await` expression in the default for the `x` parameter of `bar` is executed within the scope of `foo`. So removing the `async` on `foo` would mean that that `await` expression would no longer be found in an async context (== illegal syntax).

Reply:
Perfect. That's what this lint rule should support already today. It might be worth to copy the example as a test case.

Reply:
I've added a test case (which should pass, without causing the rule to fire) where an async function defines an inner function by setting the inner function's default argument value to the result of an await expression, below.